### PR TITLE
chore: adr for caching implementation

### PR DESCRIPTION
**Pull request by decleaver**
## Description

ADR

## Related Issue

Relates to #653 

## Type of change

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Other (security config, docs update, etc)

## Checklist before merging

- [ ] Test, docs, adr added or updated as needed
- [ ] [Contributor Guide Steps](https://github.com/defenseunicorns/uds-cli/blob/main/CONTRIBUTING.md) followed

### CODE REVIEW

File: adr/0007-caching.md [hunk 1]
```diff
@@ -0,0 +1,39 @@
+# 7. Caching
+
+Date: 17 June 2024
+
+## Status
+IN PROGRESS
+
+## Context
+UDS interacts with both local and remote bundles that are made up of both local and remote zarf packages. When dealing with remote artifacts, the process can be slow due to the time it takes to download the artifacts. We want to speed up that process by caching the artifacts. There are different ways to cache these artifacts, and we need to decide which one to use. We want to take into consideration the following: efficiency, ease of implementation, and code readability. We want to be able to leverage the same caching  implementation everywhere it makes sense. We also want the caching implementation to be resilient and able to recover from any corruption.
```
Comment (UncleGedd) [inline adr/0007-caching.md:9]:
```suggestion
UDS interacts with both local and remote bundles that are made up of both local and remote zarf packages. When dealing with remote artifacts, the process can be slow due to the time it takes to download the artifacts. We want to speed up that process by caching the layers of the artifacts. There are different ways to cache these artifacts, and we need to decide which one to use. We want to take into consideration the following: efficiency, ease of implementation, and code readability. We want to be able to leverage the same caching  implementation everywhere it makes sense. We also want the caching implementation to be resilient and able to recover from any corruption.
```

File: adr/0007-caching.md [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+
+## Options
+
+### ORAS Content-Addressable Storage (CAS) as Cache
+By leveraging how the oras-go library copies a rooted directed acyclic graph (DAG) from the source CAS to the destination CAS we can treat the destination CAS as a local storage cache if we make that destination CAS persistent. While traversing the graph, for each node, it checks if it already exists in the destination. If so, it skips the node. If the node doesn't exist in the destination, it finds the successors of the node. If the node has successors, it processes the successors first. After all successors are processed, it checks if the node exists in the in-memory cache. If so, it copies the node from the in-memory cache to the destination. If not, it copies the node from the source to the destination.
```
Comment (UncleGedd) [inline adr/0007-caching.md:14]:
```suggestion
By leveraging how the oras-go library copies a rooted directed acyclic graph (DAG) from the remote source CAS to the destination CAS we can treat the destination CAS as a local storage cache if we make that destination CAS persistent. While traversing the graph, for each node, it checks if it already exists in the destination. If so, it skips the node. If the node doesn't exist in the destination, it finds the successors of the node. If the node has successors, it processes the successors first. After all successors are processed, it checks if the node exists in the in-memory cache. If so, it copies the node from the in-memory cache to the destination. If not, it copies the node from the remote source to the destination.
```

Comment (UncleGedd) [inline adr/0007-caching.md:14]: Can we drop a link to oras-go?

**Comment by UncleGedd** [comment]
Awesome! Couple of suggestions:
- Add some context for the current state of our caching (ie. how we're using both options right now)
- Describe the chosen solution is more detail; discuss the `.uds-cache` dir and what layers we will/won't cache and why

**Review by UncleGedd — DISMISSED** [review]

**Review by TristanHoladay — DISMISSED** [review]

**Pull request merged**